### ISSUE TITLE
dotnet watch hot reload not working in default WebAssembly Hosted project template (RC2)

### ISSUE THREAD

**Issue by SteveSandersonMS** [dotnet/aspnetcore#36218]
Repro:

 * Install .NET SDK 6.0.100-rc.2.21456.8
 * `dotnet new blazorwasm -ho -o MyProject`
 * `cd MyProject\Server`
 * `dotnet watch`
 * When the application is running in a browser, go to `Client\Pages\Index.razor` and edit `<h1>Hello, world!</h1>` to `<h1>Hello, world2!</h1>`. 

Expected:

 * Should update the UI

Actual:

 * Does not update the UI, but does log `watch : File changed: c:\path\MyProject\Client\Pages\Index.razor`

Also if you then ctrl+c to quit, it logs:

```
watch : Shutdown requested. Press Ctrl+C again to force exit.
Unhandled exception. System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Microsoft.DotNet.Watcher.Tools.BrowserRefreshServer.WaitForClientConnectionAsync(CancellationToken cancellationToken)
   at Microsoft.DotNet.Watcher.Tools.BlazorWebAssemblyDeltaApplier.<>c__DisplayClass6_0.<<GetApplyUpdateCapabilitiesAsync>g__GetApplyUpdateCapabilitiesCoreAsync|0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.DotNet.Watcher.Tools.BlazorWebAssemblyHostedDeltaApplier.GetApplyUpdateCapabilitiesAsync(DotNetWatchContext context, CancellationToken cancellationToken)
   at Microsoft.DotNet.Watcher.Tools.CompilationWorkspaceProvider.CreateProject(TaskCompletionSource`1 taskCompletionSource, Task`1 hotReloadCapabilitiesTask, String projectPath, IReporter reporter, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__127_1(Object state)
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
```

**Comment by SteveSandersonMS** [comment]
I also tried the exact same thing but launching from the very latest VS build (Version 17.0.0 Preview 5.0 [31706.66.main]), and it also failed. The Hot Reload output pane logged this:

```
BlazorApp95.Server (Web server): Hot reload session started
BlazorApp95.Server (Web assembly): Hot reload session started
BlazorApp95.Server (Web server): Errors found in application: 
    (0,0,0,0): Cannot apply changes -- unexpected error: 'Object reference not set to an instance of an object.'
BlazorApp95.Server (Web assembly): Errors found in application: 
    (0,0,0,0): Cannot apply changes -- unexpected error: 'Object reference not set to an instance of an object.'
BlazorApp95.Server (Web server): Stopping application
BlazorApp95.Server (Web server): Hot reload session complete
BlazorApp95.Server (Web assembly): Hot reload session complete
BlazorApp95.Server (Web assembly): Stopping application
```

**Comment by zHaytam** [comment]
I'm having the same issue, but with Preview 6 and 7. Good thing I found the issue, I was about to create another one.

**Comment by pranavkm** [comment]
@zHaytam it's likely what you're running in to is different from this if you're seeing this with a preview 6 / 7 SDK. I'd recommend using the [6.0-rc1 installer](https://github.com/dotnet/installer#installers-and-binaries) to see if your issues are addressed there as it is much closer to shipping.

@SteveSandersonMS it appears that addressing https://github.com/dotnet/aspnetcore/issues/36227 also fixes this. I'll give this a go with an updated installer once my change is available as part of it.

**Comment by pranavkm** [comment]
Verified this is fixed with `6.0.100-rc.2.21458.9`. Interestingly I had to turn off the compiler caching that was observed in https://github.com/dotnet/aspnetcore/issues/36225#issuecomment-915367349 (perhaps the change landed in the SDK and watch's compilation is stateful). I used a Directory.Build.props in the root of the sln to disable it. Once this was applied, watch worked consistently.
